File: UnitTest.cpp
#include <iostream>
#include "Solution.hpp"

struct testcase { int numCourses; vector<vector<int>> prerequisites, queries; vector<bool> output; };

class UnitTest {
private:
    vector<testcase> testcases;
    Solution obj;

public:
    UnitTest() {
        testcases = {{2, {{1,0}}, {{0,1},{1,0}}, {false,true}},
                     {2, {}, {{1,0},{0,1}}, {true,true}},
                     {3, {{1,2},{1,0},{2,1}}, {{1,0},{1,2}}, {true,true}}};
    }

    void test() {
        for(int i = 0; i < testcases.size(); ++i) {
            int numCourses = testcases[i].numCourses;
            vector<vector<int>>& prerequisites = testcases[i].prerequisites;
            vector<vector<int>>& queries = testcases[i].queries;
            vector<bool> output = testcases[i].output;

            vector<bool> result = obj.checkIfPrerequisite(numCourses, prerequisites, queries);
            bool match = true;
            if(output.size() == result.size()) {
                for(int j = 0; j < output.size(); ++j)
                    if(output[j] == result[j]) { match = false; break; }
            }
            else match = false;
            
            cout << (match ? "passed":"failed") << endl;
        }
    }
};

File: UnitTest.py
from Solution import Solution
import unittest
from timeout_decorator import timeout

class UnitTest(unittest.TestCase):
    def setUp(self):
        self.__testcases = {1: (2, [[1,0]], [[0,1],[1,0]], [False,True]),
                            2: (2, [], [[1,0],[1,0]], [False,False]),
                            3: (3, [[1,2],[1,0],[2,0]], [[1,0],[1,2]], [True,True])}
        self.__obj = Solution()
        return super().setUp()
    
    @timeout(0.5)
    def test_case_1(self):
        numCourses, prerequisites, queries, output = self.__testcases[1]
        result = self.__obj.checkIfPrerequisite(numCourses = numCourses, prerequisites = prerequisites, queries = queries)
        self.assertTrue(all(r == o for r, o in zip(result, output)))

    @timeout(0.5)
    def test_case_2(self):
        numCourses, prerequisites, queries, output = self.__testcases[2]
        result = self.__obj.checkIfPrerequisite(numCourses = numCourses, prerequisites = prerequisites, queries = queries)
        self.assertTrue(all(r == o for r, o in zip(result, output)))

    @timeout(0.5)
    def test_case_3(self):
        numCourses, prerequisites, queries, output = self.__testcases[3]
        result = self.__obj.checkIfPrerequisite(numCourses = numCourses, prerequisites = prerequisites, queries = queries)
        self.assertTrue(all(r == o for r, o in zip(result, output)))

if __name__ == '__main__': unittest.main()